Lesson8

Hanna-Konovalova opened the pull request:
homework complete for lesson8

Hanna-Konovalova committed:
lesson8 Homework complete

Hanna-Konovalova committed:
lesson8 Homework complete

RomanKashenok commented on src/com/konovalova/lesson8/Bankomat.java:
```diff
@@ -0,0 +1,79 @@
+        this.nom100 = nom100;
+    };
+    public void putMoney(int qnom20, int qnom50, int qnom100){
+        this.nom20 += qnom20;
```
в методе getRestOfMoney ты используешь номиналы, умножая их на них же (nom20 * 20).

Hanna-Konovalova replied:
Не совсем, nom20 - это количество купюр номиналом 20. Я умножаю количество купюр на номинальную стоимость купюры.

RomanKashenok commented on src/com/konovalova/lesson8/Bankomat.java:
```diff
@@ -0,0 +1,79 @@
+            return false;
+        }
+    }
+    private boolean toMakeWithdrou(int Amount){
```
Amount – с маленькой буквы

RomanKashenok commented on src/com/konovalova/lesson8/HomeworkTask11.java:
```diff
@@ -0,0 +1,35 @@
+        System.out.println("Print Your text: ");
+        Scanner in = new Scanner(System.in);
+        String ourString = in.nextLine();
+        int threePointsIndex;
```
вместо этого лучше использовать регулярку \\w- выделишь чисто слова

RomanKashenok commented on src/com/konovalova/lesson8/HomeworkTask11.java:
```diff
@@ -0,0 +1,35 @@
+        StringBuilder stringBuilder = new StringBuilder(ourString);
+
+        while(true){
+            threePointsIndex = stringBuilder.indexOf("...");
```
на каждой итерации ты создаешь нрвую строку "..." - если ты ей постоянно пользуешься - выведи ей в константы

RomanKashenok commented on src/com/konovalova/lesson8/HomeworkTask11.java:
```diff
@@ -0,0 +1,35 @@
+            if(threePointsIndex<0){
+                break;
+            }
+            stringBuilder.replace(threePointsIndex, (threePointsIndex+3), "?");
```
не совсем понимаю смысл такой замены. почему на "?"

Hanna-Konovalova replied:
Был выбран любой знак препинания, состоящий из единичного символа для замены.

RomanKashenok commented on src/com/konovalova/lesson8/HomeworkTask12.java:
```diff
@@ -0,0 +1,35 @@
+        append, а также аналогино для StringBuffer. Сравните скорость их выполнения.
+         **/
+        String simpleString = "";
+        StringBuilder bldString = new StringBuilder(1000);
```
а зачем тут по тысяче?

RomanKashenok commented on src/com/konovalova/lesson8/HomeworkTask12.java:
```diff
@@ -0,0 +1,35 @@
+
+        //1
+        long time = System.currentTimeMillis();
+        while(simpleString.length()<1001){
```
сделай просто 100000 итераций, без while

RomanKashenok commented on src/com/konovalova/lesson8/HomeworkTask13.java:
```diff
@@ -0,0 +1,25 @@
+        Scanner in = new Scanner(System.in);
+        String ourString = in.nextLine();
+
+        Pattern regexp =Pattern.compile("[A-F0-9]+");
```
паттерн найдет все, что в верхнем регистре будет. Посмотри как записываются  шестнадцатеричные числа

RomanKashenok commented on src/com/konovalova/lesson8/HomeworkTask14.java:
```diff
@@ -0,0 +1,34 @@
+                " lacus habitasse egestas. Lobortis ipsum sapien phasellus sagittis suspendisse. Blandit commodo " +
+                " amet cubilia augue iaculis sit vestibulum imperdiet ligula posuere varius. Arcu hac risus penatibus.<p id=”p1”>";
+
+        ourString = ourString.replaceAll("<p id=.{4,5}>", "<p>");
```
у этого тега может быть гораздо больше параметров id, style, color, margin и еще штук 30 разных

RomanKashenok commented on src/com/konovalova/lesson8/HomeworkTask9.java:
```diff
@@ -0,0 +1,27 @@
+        System.out.println("Print Your text: ");
+        Scanner in = new Scanner(System.in);
+        String ourString = in.nextLine();
+        int threePointsIndex;
```
аналогично предыдущему

RomanKashenok commented on src/com/konovalova/lesson8/TimeDescription.java:
```diff
@@ -0,0 +1,81 @@
+     два конструктора: получающий общее количество секунд, и часы, минуты и секунды
+     по отдельности. Сделать метод для вывода данных.
+       **/
+
```
не нашел метода, который возвращает время в секундах
при сравнении я ожидаю, что передам в метод 2 объекта TimeDescription и получу на выходе результат сравнения - у тебя такого нет

RomanKashenok commented on src/com/konovalova/lesson8/TimeDescription.java:
```diff
@@ -0,0 +1,81 @@
+        return hours;
+    }
+
+    public int secondInTime(int hours, int minutes, int seconds){
```
непонятно к чему относится метод - он просто считает в секундах переданные ему промежутки времени (переданные кем угодно и откуда угодно) - не нужен такой

RomanKashenok commented on src/com/konovalova/lesson8/TimeDescription.java:
```diff
@@ -0,0 +1,81 @@
+          return seconds + minutes*60 + hours*60*60;
+    }
+
+    public TimeDescription fullTimeFromSeconds(int seconds){
```
для чего этого метод? не вижу смысла в нем

RomanKashenok commented on src/com/konovalova/lesson8/TimeDescription.java:
```diff
@@ -0,0 +1,81 @@
+        return new TimeDescription (ourHours,ourMinutes,ourSeconds);
+    }
+
+    public int compareTimeInSeconds(int secOne, int secTwo){
```
сравнивать надо не секунды (передавать в метод), а TimeDescription

RomanKashenok commented on src/com/konovalova/lesson8/TimeDescription.java:
```diff
@@ -0,0 +1,81 @@
+    }
+
+    public int compareFullTime(TimeDescription timeOne, TimeDescription timeTwo){
+        int compH = Integer.compare(timeOne.getHours(),timeTwo.getHours());
```
вот тут лучше все время перевести в секунды и сравнить просто два лонга

RomanKashenok commented on src/com/konovalova/lesson8/TimeDescription.java:
```diff
@@ -0,0 +1,81 @@
+        return "hours: " + this.hours + ", minutes: " + this.minutes + ", seconds: " + this.seconds;
+    }
+
+    public void printCompResInSeconds(int secOne, int secTwo){
+        int compS = compareTimeInSeconds(secOne,secTwo);
+        String ourCS = compS==0?" = ":compS==1?" > ": " < ";
+        System.out.println("Time1" + ourCS + "Time2");
+    }
+
+    public void printCompResFullTime(TimeDescription timeOne, TimeDescription timeTwo){
+        int compS = compareFullTime(timeOne,timeTwo);
+        String ourCS = compS==0?" = ":compS==1?" > ": " < ";
+        System.out.println("Full time compare\nTime1" + ourCS + "Time2");
```
странные методы. я бы их убрал

RomanKashenok commented on src/com/konovalova/lesson8/TimeDescription.java:
```diff
@@ -0,0 +1,81 @@
+        this.seconds = seconds;
+    }
+
+    public int getSeconds(){
+      return seconds;
+    }
+
+    public int getMinutes(){
+        return minutes;
+    }
+
+    public int getHours(){
+        return hours;
+    }
```
Не нужные методы, убирай

RomanKashenok commented on src/com/konovalova/lesson8/TimeDescription.java:
```diff
@@ -0,0 +1,81 @@
+    public TimeDescription(int hours, int minutes, int seconds){
+        this.hours = hours;
+        this.minutes = minutes;
+        this.seconds = seconds;
```
а если ты передашь 786567547 секунд? 
В читабельном виде было бы хорошо перевести секунды в минут, минуты в часы и добавить эти часы к переданным (вы делали похожее задание уже)

RomanKashenok replied:
ладно. А если будет с часами такая же беда? 
Я к тому, что тут должно быть все читаемо и очевидно - часов может быть сколько угодно, но минут по итогу - небольше 60, секунд - не больше 60. А если у тебя хранится 46558 часов 987656789 минут и 9876545667686 секунд, то хз сколько это. Но если при создании в конструкторе все сделать нормально, то будет хранится 26  секунд, 10 минут и ХХХ часов...

RomanKashenok commented on src/com/konovalova/lesson8/TimeDescriptionRunner.java:
```diff
@@ -0,0 +1,28 @@
+public class TimeDescriptionRunner {
+    public static void main(String[] args) {
+
+        TimeDescription timeOne = new TimeDescription(24560);
+        TimeDescription timeTwo = new TimeDescription(5,12,35);
+
+        System.out.println("Time1: " + timeOne.toString());
+        System.out.println("Time2: " + timeTwo.toString());
+
+        System.out.println("Time1 in full time format: " + timeOne.fullTimeFromSeconds(timeOne.getSeconds()).toString());
+        int newTimeInSeconds = timeTwo.secondInTime(timeTwo.getHours(),timeTwo.getMinutes(),timeTwo.getSeconds());
+        System.out.println("Time2 in seconds: " + newTimeInSeconds);
+
+        timeOne.printCompResInSeconds(timeOne.getSeconds(),newTimeInSeconds);
+        timeOne.printCompResFullTime(timeOne.fullTimeFromSeconds(timeOne.getSeconds()),timeTwo);
```
тут надо бы все переделать )) в соответствии с коментами выше

Hanna-Konovalova committed:
lesson8 HomeworkTask11 - optimize split method
lesson8 Bankomat - rewrite "amount" with cursive letter
lesson8 HomeworkTask9 - put "..." as a constant
lesson8 HomeworkTask12 - change 'while' loop to 'for'
lesson8 HomeworkTask13 - add lower case for search hexadecimal numbers
lesson8 HomeworkTask14 - change regexp for paragraph tags search
lesson8 TimeDescription - change algorithm for TimeDescription

RomanKashenok commented on src/com/konovalova/lesson8/HomeworkTask14.java:
```diff
@@ -17,11 +17,11 @@ public static void main(String[] args) {
                 " amet cubilia augue iaculis sit vestibulum imperdiet ligula posuere varius. Arcu hac risus penatibus.<p id=”p1”>";
 
-        ourString = ourString.replaceAll("<p id=.{4,5}>", "<p>");
+        ourString = ourString.replaceAll("<p.*?>", "<p>");
```
ок. Скоро узнаем, что у каждого тега должен быть еще и закрывающий ))) Пока -гут

RomanKashenok commented on src/com/konovalova/lesson8/TimeDescription.java:
```diff
@@ -13,69 +13,31 @@ public class TimeDescription {
-        return Integer.compare(secOne,secTwo);
-    }
+    public String compareTime(TimeDescription timeOne,TimeDescription timeTwo){
+        int compRes = Long.compare(timeOne.seconds,timeTwo.seconds);
```
у одного 10 часов 3 минуты и 5 секунд, у второго 1 час 2 минуты и 54 секунды. ПО твоей логике второе будет больше
